src/cargo/sources/git/utils.rs

```diff
@@ -919,47 +944,64 @@ fn init(path: &Path, bare: bool) -> CargoResult<git2::Repository> {
 /// made.
 ///
 /// This function will attempt to hit that fast path and verify that the `oid`
-/// is actually the current `master` branch of the repository. If `true` is
-/// returned then no update needs to be performed, but if `false` is returned
-/// then the standard update logic still needs to happen.
+/// is actually the current branch of the repository. If `true` is returned then
+/// no update needs to be performed, but if `false` is returned then the
+/// standard update logic still needs to happen.
 ///
 /// [1]: https://developer.github.com/v3/repos/commits/#get-the-sha-1-of-a-commit-reference
 ///
 /// Note that this function should never cause an actual failure because it's
 /// just a fast path. As a result all errors are ignored in this function and we
 /// just return a `bool`. Any real errors will be reported through the normal
 /// update path above.
-fn github_up_to_date(handle: &mut Easy, url: &Url, oid: &git2::Oid) -> bool {
-    macro_rules! r#try {
-        ($e:expr) => {
-            match $e {
-                Some(e) => e,
-                None => return false,
-            }
-        };
-    }
+fn github_up_to_date(
+    repo: &mut git2::Repository,
+    url: &str,
+    reference: &GitReference,
+    config: &Config,
+) -> CargoResult<bool> {
+    let url = Url::parse(url)?;
+    if url.host_str() != Some("github.com") {
+        return Ok(false);
+    }
+
+    let github_branch_name = match reference {
+        GitReference::Branch(branch) => branch,
+        GitReference::Tag(tag) => tag,
+        GitReference::Rev(_) => {
+            debug!("can't use github fast path with `rev`");
+            return Ok(false);
+        }
+    };
 
     // This expects GitHub urls in the form `github.com/user/repo` and nothing
     // else
-    let mut pieces = r#try!(url.path_segments());
-    let username = r#try!(pieces.next());
-    let repo = r#try!(pieces.next());
+    let mut pieces = url
+        .path_segments()
+        .ok_or_else(|| anyhow!("no path segments on url"))?;
+    let username = pieces
+        .next()
+        .ok_or_else(|| anyhow!("couldn't find username"))?;
+    let repository = pieces
+        .next()
+        .ok_or_else(|| anyhow!("couldn't find username"))?;
```

Comment:
```suggestion
        .ok_or_else(|| anyhow!("couldn't find repository name"))?;
```

Reply:
One thing I've noticed is that this doesn't work if the URL has `.git` on the end (like https://github.com/rust-lang/cargo.git).  Maybe it should strip `.git` if it is in the URL?